utils/zerovec/src/map/hashmap.rs

```diff
@@ -0,0 +1,270 @@
+            let l1 = bucket_flatten[start].0 as usize;
+            #[allow(clippy::indexing_slicing)] // l1 is always within bounds of `bucket_sizes`
+            let end = start + bucket_sizes[l1];
+            #[allow(clippy::indexing_slicing)] // start, end are always within bounds of
+            // `bucket_sizes`
+            let buckets = &bucket_flatten[start..end];
+
+            'seed: for seed in 0x1u32.. {
+                current_displacements.clear();
+                generation += 1;
+
+                for (_, (k, _)) in buckets {
+                    let displacement_idx = h(seed, k.borrow(), iter_len) as usize;
+                    #[allow(clippy::indexing_slicing)] // displacement_idx is always within bounds
+                    if occupied[displacement_idx] || assignments[displacement_idx] == generation {
+                        continue 'seed;
+                    }
+                    current_displacements.push(displacement_idx);
+                    if let Some(v) = assignments.get_mut(displacement_idx) {
+                        *v = generation;
+                    }
+                }
+
+                // Successfully found a seed, store it as index l1.
+                if let Some(v) = displacements.get_mut(l1) {
+                    *v = seed;
+                }
+
+                for (i, displacement_idx) in current_displacements.iter().enumerate() {
+                    #[allow(clippy::indexing_slicing)] // `current_displacements` has same size as
+                    // `buckets`
+                    let (_, (_, original_idx)) = &buckets[i];
+
+                    // Make clippy happy
+                    if let Some(v) = occupied.get_mut(*displacement_idx) {
+                        *v = true;
+                    }
+                    if let Some(v) = reverse_mapping.get_mut(*displacement_idx) {
+                        *v = *original_idx as u32;
+                    }
+                }
+                break;
+            }
+            start = end;
+        }
+
+        Self {
+            displacements: ZeroVec::alloc_from_slice(&displacements),
+            reverse_mapping: ZeroVec::alloc_from_slice(&reverse_mapping),
+        }
+    }
+
+    #[inline]
+    pub fn index<K>(&self, k: &K) -> Option<usize>
+    where
+        K: Hash,
+    {
+        let l1 = compute_hash(0, k, self.displacements.len());
+
+        #[allow(clippy::unwrap_used)] // l1 is in 0..self.displacements.len()
+        let seed = self.displacements.get(l1 as usize).unwrap();
+        if seed == 0 {
+            None
+        } else {
+            let hash = compute_hash(seed, k, self.displacements.len());
+            self.reverse_mapping.get(hash as usize).map(|i| i as usize)
+        }
+    }
+}
+
+pub struct ZeroHashMapStatic<'a, K, V>
+where
+    K: ZeroMapKV<'a> + ?Sized,
+    V: ZeroMapKV<'a> + ?Sized,
+{
+    index: HashIndex<'a>,
+    keys: K::Container,
```

Comment:
Suggestion: We don't need the keys for lookup, so we shouldn't store them. We won't be able to have `iter_keys` but that's fine. If necessary we could have a `ZeroHashMapWithKeys` if someone needs this (or just encourage them to use `ZeroMap` if they care about iterating over the keys).

Reply:
Removed keys from `ZeroHashMapStatic`.

Reply:
Huh, this is not how hashmaps usually work. What about hash collisions? We should have to compare to the actual key somewhere, unless this algorithm somehow encodes unlimited entropy in two `u32`s.

Reply:
Should we have two variants then as if we support unknown keys it would lead to a comparison of the key on every lookup.

Reply:
Long term yes, but for now we should focus on the general case, the key-less one can be an optimization for later.

Reply:
It's a discussion to be had. Given the multi-stage lookup, it seems that hash collisions are unlikely. It would be reasonable to argue that it's okay to (rarely) get GIGO results, so long as that possibility is documented, and if you really care about avoiding collisions, use ZeroMap.

Reply:
Perfect minimal hash functions mean that *every* key that is not in the map collides with some key in the map. It's not that some keys might collide with probability 2^-32 or something, the codomain of the hash function is the size of the collection so without storing the keys we will not know which keys are in the map and which aren't.

Reply:
Yes if the first hash matches you are guaranteed a collision

Reply:
OK. You're right. I should read the paper.

Reply:
Take a look at http://stevehanov.ca/blog/?id=119